python/cudf/cudf/core/dataframe.py

```diff
@@ -3375,37 +3308,37 @@ def from_pandas(cls, dataframe, nan_as_null=True):
 
         df = cls()
         # Set columns
-        for colk in dataframe.columns:
+        for i, colk in enumerate(dataframe.columns):
             vals = dataframe[colk].values
             # necessary because multi-index can return multiple
             # columns for a single key
             if len(vals.shape) == 1:
-                df[colk] = Series(vals, nan_as_null=nan_as_null)
+                df[i] = Series(vals, nan_as_null=nan_as_null)
             else:
                 vals = vals.T
                 if vals.shape[0] == 1:
-                    df[colk] = Series(vals.flatten(), nan_as_null=nan_as_null)
+                    df[i] = Series(vals.flatten(), nan_as_null=nan_as_null)
                 else:
                     # TODO fix multiple column with same name with different
                     # method.
                     if isinstance(colk, tuple):
                         colk = str(colk)
                     for idx in range(len(vals.shape)):
-                        df[colk + str(idx)] = Series(
-                            vals[idx], nan_as_null=nan_as_null
-                        )
+                        df[i] = Series(vals[idx], nan_as_null=nan_as_null)
+
+        # Set columns
+        if isinstance(dataframe.columns, pd.MultiIndex):
+            df.columns = cudf.MultiIndex.from_pandas(dataframe.columns)
```

Comment:
Isn't `df.columns` a host object anyway? We'd want it to be a Pandas object.

Reply:
As discussed offline, `df.column` is a device object for the case of `MultiIndex`. Changing this is definitely out of scope for this PR